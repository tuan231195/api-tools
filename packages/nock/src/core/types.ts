import nock, { ReplyHeaders } from 'nock';
import { OverrideOptions } from '@vdtn359/api-tools-faker';

export type ExtendedNockInterceptor = nock.Interceptor & {
	fakeReply: (
		status: number,
		overrideOptions?: OverrideOptions,
		headers?: ReplyHeaders
	) => nock.Scope;
};
