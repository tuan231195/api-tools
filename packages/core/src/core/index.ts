export * from './parser';
export * from './generator';
export * from './get';
export * from './utils';
