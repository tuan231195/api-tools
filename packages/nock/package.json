{
	"name": "@vdtn359/api-tools-nock",
	"version": "1.0.0",
	"description": "mock api with nock",
	"main": "dist/index.js",
	"types": "dist/index.d.ts",
	"files": [
		"dist"
	],
	"scripts": {
		"build": "rm -rf dist && ttsc -b tsconfig.build.json",
		"prepare": "generate-api-types axios -u ../../test/v3/petstore.json -o src/generated"
	},
	"license": "ISC",
	"dependencies": {
		"@vdtn359/api-tools-core": "workspace:*",
		"@vdtn359/api-tools-faker": "workspace:*",
		"lodash": "^4.17.21",
		"path-to-regexp": "^6.2.0",
		"nock": "^13.1.1"
	},
	"devDependencies": {
		"axios": "^0.21.1",
		"@vdtn359/api-tools-types-generator": "workspace:*"
	}
}
