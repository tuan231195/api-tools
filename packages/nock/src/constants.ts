export const NAMESPACE = 'api-tools:nock';
