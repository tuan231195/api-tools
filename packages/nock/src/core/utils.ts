export function wrap<W extends object, B>(wrapper: W, base: B): W & B {
	const proxy = new Proxy(wrapper, {
		get(target, property) {
			if (property in target) {
				return getProperty(target, property, proxy);
			}

			return getProperty(base, property, proxy);
		},
	}) as W & B;
	return proxy;
}

export function getProperty(
	object: any,
	property: string | symbol,
	proxy: any
) {
	const propertyValue: any = object[property];
	if (typeof propertyValue === 'function' && !propertyValue.__proxy) {
		const oldFunction = propertyValue;
		object[property] = function (...args: any[]) {
			const value = oldFunction.apply(object, args);
			return value === object ? proxy : value;
		} as any;

		propertyValue.__proxy = true;
	}
	return propertyValue;
}
