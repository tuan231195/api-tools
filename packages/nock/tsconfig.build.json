{
	"extends": "./tsconfig.json",
	"include": ["./src"],
	"exclude": ["./src/**/*.spec.ts"],
	"references": [
		{
			"path": "../core/tsconfig.build.json"
		},
		{
			"path": "../faker/tsconfig.build.json"
		}
	],
	"compilerOptions": {
		"rootDir": "./src"
	}
}
