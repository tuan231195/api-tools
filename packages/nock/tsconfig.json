{
	"extends": "../../tsconfig.json",
	"compilerOptions": {
		"composite": true,
		"baseUrl": "./src",
		"rootDir": "./src",
		"tsBuildInfoFile": "./dist/tsconfig.tsbuildinfo",
		"outDir": "./dist",
		"paths": {
			"src/*": ["*"]
		}
	},
	"include": ["src"]
}
