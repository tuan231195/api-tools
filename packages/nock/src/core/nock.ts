import {
	getOperation,
	GetOperationArgs,
	parse,
	ParsedDocument,
	ParsedOperationSchemaWithInfo,
	matchUri,
	parseUri,
	ramlToExpress,
} from '@vdtn359/api-tools-core';
import { mergeWith } from 'lodash';
import nockBase, {
	Interceptor,
	Options,
	ReplyHeaders,
	RequestBodyMatcher,
	Scope,
} from 'nock';
import { wrap } from 'src/core/utils';
import {
	ExtendedInterceptor,
	ExtendedNockInterceptor,
	MockApi,
	OverrideOptions,
} from 'src/core/types';
import { fakeOperationResponse } from '@vdtn359/api-tools-faker';

export const nock = async (baseUrl: string, uri: string): Promise<MockApi> => {
	const parsedDocument = await parse(uri);
	const scope = nockBase(baseUrl);

	return new MockApiImpl(scope, parsedDocument);
};

class MockApiImpl implements MockApi {
	constructor(public scope: Scope, private parsedDocument: ParsedDocument) {}

	mock(
		options: GetOperationArgs,
		requestBody?: RequestBodyMatcher,
		interceptorOptions?: Options
	): ExtendedNockInterceptor {
		const operation = getOperation(this.parsedDocument, options);
		const operationMethod = operation.info.method;
		const expressPathStyle = ramlToExpress(operation.info.path);
		const operationInterceptor = this.scope[operationMethod](
			matchUri(expressPathStyle),
			requestBody,
			interceptorOptions
		);

		return wrap(
			new ExtendedInterceptorImpl(operation, operationInterceptor),
			operationInterceptor
		);
	}
}

class ExtendedInterceptorImpl implements ExtendedInterceptor {
	constructor(
		private operation: ParsedOperationSchemaWithInfo,
		private interceptor: Interceptor
	) {}

	fakeReply(overrideOptions?: OverrideOptions, headers?: ReplyHeaders): Scope;
	fakeReply(
		status: number,
		overrideOptions?: OverrideOptions,
		headers?: ReplyHeaders
	): Scope;
	fakeReply(
		firstArg?: OverrideOptions | number,
		secondArg?: ReplyHeaders | OverrideOptions,
		thirdArg?: ReplyHeaders
	): Scope {
		let responseStatus: number | undefined;
		let headers: ReplyHeaders | undefined;
		let overrideOptions: OverrideOptions | undefined;
		let status: number;
		if (typeof firstArg !== 'number') {
			headers = secondArg as ReplyHeaders | undefined;
			overrideOptions = firstArg as OverrideOptions | undefined;
			status = 200;
			responseStatus = undefined;
		} else {
			headers = thirdArg as ReplyHeaders | undefined;
			overrideOptions = secondArg as OverrideOptions | undefined;
			status = firstArg;
			responseStatus = status;
		}
		return this.interceptor.reply(
			status,
			async (uri: string, body?: any) => {
				const expressPathStyle = ramlToExpress(
					this.operation.info.path
				);

				const fakeResponse = await fakeOperationResponse({
					schema: this.operation.schema,
					mergeOptions: overrideOptions?.mergeOptions,
					status: responseStatus,
				});
				let responseBody: any;
				if (typeof overrideOptions?.overrides === 'function') {
					const { params } = parseUri(uri, expressPathStyle);
					responseBody = overrideOptions.overrides({
						uri,
						body,
						fakeResponse,
						params,
					});
				} else if (overrideOptions?.overrides) {
					responseBody = mergeWith(
						fakeResponse,
						overrideOptions.overrides,
						overrideOptions.mergeOptions
					);
				}
				return responseBody;
			},
			headers
		);
	}
}
