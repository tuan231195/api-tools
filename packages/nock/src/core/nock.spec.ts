import { nock } from 'src/core/nock';
import path from 'path';
import axios from 'axios';
import { createClient, createClientWithAxiosResponse } from 'src/generated';
import { initSeed } from '@vdtn359/api-tools-faker';
import { MockApi } from 'src/core/types';

initSeed('123');

const testDir = path.resolve(__dirname, '..', '..', '..', '..', 'test');

const baseURL = 'http://test.com';

describe('nock', () => {
	let mockApi: MockApi;

	beforeEach(async () => {
		mockApi = await nock(
			baseURL,
			path.resolve(testDir, 'v3', 'petstore.json')
		);
	});

	it('should mock the response', async () => {
		const axiosInstance = axios.create({
			baseURL,
		});
		const client = createClient(axiosInstance);
		mockApi
			.mock({
				operationId: 'createUser',
			})
			.fakeReply(200, {
				overrides: {
					id: 1,
				},
			});
		const response = await client.createUser({
			body: {
				id: '1',
			} as any,
		});
		expect(response).toMatchSnapshot();
	});

	it('should allow override function', async () => {
		const axiosInstance = axios.create({
			baseURL,
		});
		const overrideFunction = jest
			.fn()
			.mockImplementation(() => mockResponse);
		const client = createClientWithAxiosResponse(axiosInstance);
		const mockResponse: any = {
			id: 1,
			username: 'blah',
		};
		const mockRequest: any = {
			id: 1,
			firstName: 'John',
			lastName: 'Doe',
		};
		mockApi
			.mock(
				{
					methodName: 'put',
					path: '/user/{username}',
				},
				mockRequest
			)
			.fakeReply(
				{
					overrides: overrideFunction,
				},
				{
					'x-header': 'test',
				}
			);
		const response = await client.updateUser({
			body: mockRequest,
			path: {
				username: mockResponse.username,
			},
		});
		expect(response.data).toEqual(mockResponse);
		expect(response.headers).toEqual({
			'x-header': 'test',
			'content-type': 'application/json',
		});
		expect(overrideFunction).toHaveBeenCalledWith({
			body: mockRequest,
			fakeResponse: expect.objectContaining({ id: expect.any(String) }),
			uri: '/user/blah',
			params: { username: 'blah' },
		});
	});
});
