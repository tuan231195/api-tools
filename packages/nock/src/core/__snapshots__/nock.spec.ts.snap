// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nock should mock the response 1`] = `
Object {
  "email": "dasinih@nadedok.ch",
  "firstName": "Jerry",
  "id": 1,
  "lastName": "Simoni",
  "password": "RNVhUantPjOu",
  "phone": "(526) 893-4707",
  "userStatus": true,
  "username": "ZBwiHHrAoRhZiENxHDXh",
}
`;
